### ISSUE TITLE
Using built-in set/get_ctx_params functions in user's provider

### ISSUE THREAD

**Issue by alaakryeem** [openssl/openssl#19444]
When writing a provider with RSA signature algorithm, the `EVP_PKEY_CTX_set_rsa_padding` API is failing, after debugging the failure I came to a conclusion that it's failing because `OSSL_FUNC_signature_settable_ctx_params_fn, OSSL_FUNC_signature_set_ctx_params_fn` functions are missing in my OSSL_DISPATCH array. This is what I have so far:
```
static const OSSL_DISPATCH dummy_RSAsig_funcs[] = {
    { OSSL_FUNC_SIGNATURE_NEWCTX, (void (*)(void))dummy_sig_newctx },
    { OSSL_FUNC_SIGNATURE_FREECTX, (void (*)(void))dummy_sig_freectx },
    { OSSL_FUNC_SIGNATURE_SIGN_INIT, (void (*)(void))dummy_sig_sign_init },
    { OSSL_FUNC_SIGNATURE_SIGN, (void (*)(void))dummy_RSAsig_sign },

    { 0, NULL }
};
//**missing**:
//{ OSSL_FUNC_SIGNATURE_SETTABLE_CTX_PARAMS, (void (*)(void))dummy_settable_ctx_params },
//{ OSSL_FUNC_SIGNATURE_SET_CTX_PARAMS, (void (*)(void))dummy_set_ctx_params },
```
My question is: Is it possible to re-use the functions that the OpenSSL RSA implementation has? Or it's a must to implement the whole functionality in my provider?
I only want to change the signing logic and keep all other wrapping methods as is (default openssl RSA).

**Comment by paulidale** [comment]
TL;DR: don't do it.

You could reuse OpenSSL's functions if you can get access to them.  However, there are a number of problems:
* we don't make any guarantees about memory layout or format of our key material -- if anything changes, you **will** break.  _For me this is the show stopper._
* we don't make these functions public outside the provider.  _I'll leave it as an exercise as to how to get them._
* this really isn't a good idea.

A better option would be to copy the code into your provider & use it that way.

**Comment by levitte** [comment]
This isn't just about openssl's implementations, it's a design decision. Providers are meant to be self-contained.

**Comment by alaakryeem** [comment]
@paulidale , @levitte 
To make sure I fully understand this, there is no option today to customize a specific functionality in OpenSSL operations?
E.g. If I want to log each RSA signing operation that OpenSSL 3.0 performs, is there a way to do it? With Engines I could use ENGINE_set_RSA to set a custom function of mine that writes to a log and then calls the actual sign function of OpenSSL, is this possible in OpenSSL 3.0?

**Comment by t8m** [comment]
In theory, you could use ideas from the filtering provider test/filterprov.c and put your provider in between the default provider and libcrypto. So yeah, it is possible to do some customization of provider operations although it might not be as easy as with the ENGINE api and methods.

**Comment by alaakryeem** [comment]
I looked into the filtering provider, it implements a nice idea, but it still doesn't fully cover ENGINE's set_method, because you can't handle specific functionality as part of the desired operation (in my example, I wanted to _filter/handle_ the OSSL_FUNC_SIGNATURE_SIGN functionality from OSSL_OP_SIGNATURE operation and forward all other functionalities and operations to the default provider).
Also, such functionality can't be changed since the function pointers are treated as const in the OSSL__DISPATCH struct.
As I understand the idea is to let user's provider handle some operations (e.g. OSSL_OP_SIGNATURE) and forward all other operations to the default provider.

**Comment by alaakryeem** [comment]
Hi @t8m,
Do you have any suggestions here?

**Comment by t8m** [comment]
I am afraid I do not have any better suggestions. The provider API is not a direct replacement of the engine/method API and one of the reasons for deprecating the engine/method APIs was that this exposes too much of the OpenSSL internals. So I am afraid you'll have to add these methods to your provider yourself.
